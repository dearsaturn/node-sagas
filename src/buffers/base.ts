import {IActionBuffer, IAction} from '../types';

export default class BaseBuffer<Action extends IAction> implements IActionBuffer<Action> {
    protected actions: Action[] = [];
    protected observers: Array<typeof Promise['resolve']> = [];

    public isEmpty() {
        return this.actions.length === 0;
    }

    public put(action: Action) {
        this.actions = [...this.actions, action];
        this.notifyObservers();
    }

    public notifyObservers() {
        throw new Error('notifyObservers is not implemented in the base class');
    }

    public async take(): Promise<Action> {
        let resolve;
        const promise = new Promise(r => (resolve = r));
        if (resolve) {
            this.observers.push(resolve);
        }
        return promise as Promise<Action>;
    }
}
