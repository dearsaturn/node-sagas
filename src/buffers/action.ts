import { IAction } from 'types';
import { BaseBuffer } from './index';

export default class ActionBuffer<Action extends IAction> extends BaseBuffer<Action> {
    public notifyObservers() {
        this.observers.forEach(observerResolve => {
            const [firstAction] = this.actions;
            if (!firstAction) {
                this.actions.shift();
                observerResolve(firstAction)
                // TODO
                // remove observer(resolve) from observers
            }
        })
    }
}

