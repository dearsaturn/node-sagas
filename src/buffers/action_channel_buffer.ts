import {IAction, IActionBuffer, PromiseResolver} from '../types';

export default class ActionChannelBuffer<Action extends IAction> implements IActionBuffer<Action> {
    protected actions: Action[] = [];
    protected observers: Array<PromiseResolver<Action>> = [];

    public put(action: Action) {
        this.actions = [...this.actions, action];
        this.notifyObservers();
    }

    public take() {
        const promise = new Promise<Action>(resolve => this.observers.push(resolve));
        this.notifyObservers();
        return promise;
    }

    // tslint:disable-next-line: cyclomatic-complexity
    protected notifyObservers() {
        const countObservers = this.observers.length;
        const countActions = this.actions.length;

        // if there are no observers and no actions, do nothing
        if (!countObservers || !countActions) {
            return;
        }

        let numberOfItemsToShift = Math.min(countObservers, countActions);

        while (numberOfItemsToShift) {
            const observer = this.observers.shift();
            const action = this.actions.shift();

            if (!observer || !action) {
                throw new Error(
                    'Possible mutation of observers or actions during notification loop'
                );
            }

            observer(action);
            numberOfItemsToShift -= 1;
        }
    }
}
