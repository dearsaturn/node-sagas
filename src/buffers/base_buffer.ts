import {IActionBuffer, IAction, PromiseResolver} from '../types';

export default class BaseBuffer<Action extends IAction> implements IActionBuffer<Action> {
    protected actions: Action[] = [];
    protected observers: Array<PromiseResolver<Action>> = [];

    public isEmpty() {
        return this.actions.length === 0;
    }

    public put(action: Action) {
        this.actions = [...this.actions, action];
        this.notifyObservers();
    }

    public notifyObservers() {
        throw new Error('notifyObservers is not implemented in the base class');
    }

    public take(): Promise<Action> {
        const promise = new Promise<Action>(resolve => this.observers.push(resolve));
        this.notifyObservers();
        return promise;
    }
}
