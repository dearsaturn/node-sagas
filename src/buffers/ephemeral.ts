import {BaseBuffer} from './index';
import {IAction} from '../types';

export default class EphemeralBuffer<Action extends IAction> extends BaseBuffer<Action> {
    public notifyObservers() {
        this.observers.forEach(observerResolve => {
            const [firstAction] = this.actions;
            if (!firstAction) {
                this.actions.shift();
                observerResolve(firstAction);
            }
        });
        this.observers = [];
    }
}
