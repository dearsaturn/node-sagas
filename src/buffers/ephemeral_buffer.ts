import {IAction, PromiseResolver, IActionBuffer} from '../types';

export default class EphemeralBuffer<Action extends IAction> implements IActionBuffer<Action> {
    protected observers: Array<PromiseResolver<Action>> = [];

    public put(action: Action) {
        for (const observerResolve of this.observers) {
            observerResolve(action);
        }

        delete this.observers;
        this.observers = [];
    }

    public take(): Promise<Action> {
        const promise = new Promise<Action>(resolve => this.observers.push(resolve));
        return promise;
    }
}
