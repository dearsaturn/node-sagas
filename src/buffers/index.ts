export {default as ActionChannelBuffer} from './action_channel_buffer';
export {default as EphemeralBuffer} from './ephemeral_buffer';
