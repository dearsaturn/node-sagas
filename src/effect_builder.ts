import {
    PutEffect,
    TakeEffect,
    CallEffect,
    ActionChannel,
    IAction,
    IActionBuffer,
    TakePattern
} from 'types';

export default function effectBuilder(transactionId: string) {
    const put = <Payload>(
        ...args: Parameters<PutEffect<Payload>>
    ): ReturnType<PutEffect<Payload>> => ({
        pattern: args[0],
        payload: args[1],
        transactionId,
        kind: 'PUT'
    });

    const take: TakeEffect = <Patterns = TakePattern>(patterns: Patterns) => ({
        transactionId,
        patterns,
        kind: 'TAKE'
    });

    const call = <Fn extends (...args: any[]) => any>(
        effect: Fn,
        args: Parameters<Fn>
    ): ReturnType<CallEffect<Fn>> => ({
        transactionId,
        kind: 'CALL',
        effect,
        args
    });

    const actionChannel = <Action extends IAction>(
        pattern: Parameters<ActionChannel<Action>>[0],
        actionBuffer?: Parameters<ActionChannel<Action>>[1]
    ): ReturnType<ActionChannel<Action>> => {
        const defaultActionBuffer: IActionBuffer<Action> = {
            isEmpty: () => true,
            put: action => action,
            take: () => ({transactionId, topic: 'stddfn'} as Action)
        };

        const buffer = actionBuffer || defaultActionBuffer;

        return {
            transactionId,
            pattern,
            buffer,
            kind: 'ACTION_CHANNEL'
        };
    };

    return {
        put,
        take,
        call,
        actionChannel
    };
}
