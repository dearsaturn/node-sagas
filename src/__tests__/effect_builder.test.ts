import {EffectBuilder} from 'effect_builder';
import {EphemeralBuffer, ActionChannelBuffer} from 'buffers';

// tslint:disable-next-line: no-empty
describe(EffectBuilder.name, function() {
    const builder = new EffectBuilder('test-transaction-id');

    describe('#put', function() {
        it('returns a put effect description', function() {
            expect(builder.put('pattern')).toMatchSnapshot();
            expect(builder.put('pattern', {something: true})).toMatchSnapshot();
        });
    });

    describe('#take', function() {
        describe('given simple patterns', function() {
            it('returns a take effect description with an ephemeral buffer', function() {
                const {buffer, ...rest} = builder.take('test');
                expect(rest).toMatchSnapshot();
                expect(buffer).toBeInstanceOf(EphemeralBuffer);
            });
        });

        describe('given an action channel pattern', function() {
            it("returns a take effect description with the action channel's buffer", function() {
                const {buffer, ...rest} = builder.take<undefined>(builder.actionChannel('plemp'));
                expect(rest).toMatchSnapshot();
                expect(buffer).toBeInstanceOf(ActionChannelBuffer);
            });
        });
    });

    describe('#call', function() {
        it('returns a call effect description', function() {
            expect(
                builder.callFn(() => {
                    return 3;
                }, [])
            ).toMatchSnapshot();
        });
    });

    describe('#actionChannel', function() {
        it('returns an action channel effect description', function() {
            expect(builder.actionChannel('smart')).toMatchSnapshot();
            expect(builder.actionChannel(['smart', 'dumb'])).toMatchSnapshot();
        });
    });

    describe('#all', function() {
        it('returns an all effect combinator description', function() {
            expect(builder.all([builder.put('asf'), builder.put('bsf')])).toMatchSnapshot();

            expect(
                builder.all({
                    a: builder.put('asf'),
                    b: builder.put('asf')
                })
            ).toMatchSnapshot();
        });
    });

    describe('#race', function() {
        it('returns a race effect combinator description', function() {
            expect(builder.race([builder.put('asf'), builder.put('bsf')])).toMatchSnapshot();

            expect(
                builder.race({
                    a: builder.put('asf'),
                    b: builder.put('asf')
                })
            ).toMatchSnapshot();
        });
    });
});
