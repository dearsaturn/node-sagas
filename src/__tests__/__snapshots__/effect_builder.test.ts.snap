// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EffectBuilder #actionChannel returns an action channel effect description 1`] = `
Object {
  "buffer": ActionChannelBuffer {
    "actions": Array [],
    "observers": Array [],
  },
  "kind": "ACTION_CHANNEL",
  "observer": [Function],
  "pattern": "smart",
  "topics": Array [
    "smart",
  ],
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #actionChannel returns an action channel effect description 2`] = `
Object {
  "buffer": ActionChannelBuffer {
    "actions": Array [],
    "observers": Array [],
  },
  "kind": "ACTION_CHANNEL",
  "observer": [Function],
  "pattern": Array [
    "smart",
    "dumb",
  ],
  "topics": Array [
    "smart",
    "dumb",
  ],
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #all returns an all effect combinator description 1`] = `
Object {
  "combinator": [Function],
  "effects": Array [
    Object {
      "kind": "PUT",
      "pattern": "asf",
      "payload": undefined,
      "topic": "asf",
      "transactionId": "test-transaction-id",
    },
    Object {
      "kind": "PUT",
      "pattern": "bsf",
      "payload": undefined,
      "topic": "bsf",
      "transactionId": "test-transaction-id",
    },
  ],
  "kind": "COMBINATOR",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #all returns an all effect combinator description 2`] = `
Object {
  "combinator": [Function],
  "effects": Object {
    "a": Object {
      "kind": "PUT",
      "pattern": "asf",
      "payload": undefined,
      "topic": "asf",
      "transactionId": "test-transaction-id",
    },
    "b": Object {
      "kind": "PUT",
      "pattern": "asf",
      "payload": undefined,
      "topic": "asf",
      "transactionId": "test-transaction-id",
    },
  },
  "kind": "COMBINATOR",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #call returns a call effect description 1`] = `
Object {
  "args": Array [],
  "effect": [Function],
  "kind": "CALL",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #put returns a put effect description 1`] = `
Object {
  "kind": "PUT",
  "pattern": "pattern",
  "payload": undefined,
  "topic": "pattern",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #put returns a put effect description 2`] = `
Object {
  "kind": "PUT",
  "pattern": "pattern",
  "payload": Object {
    "something": true,
  },
  "topic": "pattern",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #race returns a race effect combinator description 1`] = `
Object {
  "combinator": [Function],
  "effects": Array [
    Object {
      "kind": "PUT",
      "pattern": "asf",
      "payload": undefined,
      "topic": "asf",
      "transactionId": "test-transaction-id",
    },
    Object {
      "kind": "PUT",
      "pattern": "bsf",
      "payload": undefined,
      "topic": "bsf",
      "transactionId": "test-transaction-id",
    },
  ],
  "kind": "COMBINATOR",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #race returns a race effect combinator description 2`] = `
Object {
  "combinator": [Function],
  "effects": Object {
    "a": Object {
      "kind": "PUT",
      "pattern": "asf",
      "payload": undefined,
      "topic": "asf",
      "transactionId": "test-transaction-id",
    },
    "b": Object {
      "kind": "PUT",
      "pattern": "asf",
      "payload": undefined,
      "topic": "asf",
      "transactionId": "test-transaction-id",
    },
  },
  "kind": "COMBINATOR",
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #take given an action channel pattern returns a take effect description with the action channel's buffer 1`] = `
Object {
  "kind": "TAKE_ACTION_CHANNEL",
  "observer": [Function],
  "patterns": "plemp",
  "topics": Array [
    "plemp",
  ],
  "transactionId": "test-transaction-id",
}
`;

exports[`EffectBuilder #take given simple patterns returns a take effect description with an ephemeral buffer 1`] = `
Object {
  "kind": "TAKE",
  "observer": [Function],
  "patterns": "test",
  "topics": Array [
    "test",
  ],
  "transactionId": "test-transaction-id",
}
`;
